src/lib.rs

```diff
@@ -0,0 +1,742 @@
+pub mod errors;
+pub mod file_watcher;
+pub mod systeminfo_watcher;
+
+use crate::errors::TelemetryError;
+
+use base64::{engine::general_purpose::STANDARD, Engine};
+use chrono::Utc;
+use dirs::home_dir;
+use nix::{
+    errno::Errno,
+    fcntl::{Flock, FlockArg},
+    sys::stat,
+    unistd::{chdir, close, dup2, fork, setsid, sysconf, SysconfVar},
+};
+use regex::Regex;
+use std::{
+    env::{var, var_os},
+    fs::{create_dir_all, File, OpenOptions},
+    io::{stderr, stdout, Write},
+    os::fd::AsRawFd,
+    path::{Path, PathBuf},
+    process::exit,
+    sync::LazyLock,
+};
+use sysinfo::System;
+use tracing::{
+    span::{Entered, Id, Record},
+    Event, Span, Subscriber,
+};
+use tracing_appender::non_blocking::{NonBlocking, WorkerGuard};
+use tracing_subscriber::layer::Context;
+use tracing_subscriber::{
+    fmt::{
+        format::{DefaultFields, FormatEvent, FormatFields, Writer},
+        FmtContext, FormattedFields, Layer,
+    },
+    layer::SubscriberExt,
+    registry::LookupSpan,
+    util::SubscriberInitExt,
+    Layer as LayerTrait, Registry,
+};
+
+pub type Result<T> = std::result::Result<T, TelemetryError>;
+
+/// Define a prelude to minimise a simple API
+///
+/// Exports the following names:
+///
+/// - `telemetry_init`
+/// - `TelemetryLayer`
+///
+/// - `trace!`
+/// - `debug!`
+/// - `info!`
+/// - `warn!`
+/// - `error!`
+///
+/// - `event!`
+/// - `span!`
+/// - `Level`
+pub mod prelude {
+    pub use crate::{
+        debug, error, event, info, span, telemetry_init, trace, warn, Level, TelemetryLayer,
+    };
+}
+
+/// Convenience function to do everything needed for telemetry in one step
+///
+/// This function does the following:
+/// - Creates a new `TelemetryLayer` `tracing` `Layer`
+/// - Sets it as the global default `tracing` `Subscriber`
```

Comment:
This sounds like it would override the subscriber from forc-tracing since there can only be one global default. Have you tested using both of them together?

Maybe instead of `telemetry_init`, this should be built in to `init_tracing_subscriber` in forc-tracing? There could be an option like telemetry: bool  that when true, initializes with the custom layer that has `.with_span_events(FmtSpan::CLOSE)` to capture duration of `#[instrument]`ed functions

That way, it would be as simple as adding the `telemetry: true` option to any tools that already use forc-tracing. What do you think?

Reply:
Yep, I've been debating with myself about if this is actually needed or not, and was more as a convenience one-liner for the `example/` apps. I still leaning on the side of it's needed, but I think I can be removed once I add the `info_telemetry!()` macros.

Also, adding this to `forc-tracing` is an excellent idea and will be doing that once this has all finalised.

Reply:
Now that we have more ergonomic constructors via macros, I've removed `init_telemetry!()`.